src/optimisers.py

```diff
@@ -0,0 +1,149 @@
+from pymoo.operators.sampling.rnd import FloatRandomSampling
+from pymoo.algorithms.soo.nonconvex.pso import PSO
+from pymoo.core.problem import ElementwiseProblem
+from pymoo.algorithms.soo.nonconvex.ga import GA
+from pymoo.termination import get_termination
+from pymoo.operators.crossover.sbx import SBX
+from pymoo.algorithms.moo.nsga2 import NSGA2
+from pymoo.operators.mutation.pm import PM
+from pymoo.optimize import minimize
+import numpy as np
+
+
+
+
+
+class Problem(ElementwiseProblem):
+    """
+    Probem class allows a function to be minimised.
+    """
+    def __init__(self, num_dim :int, num_obj :int, loss_function :callable, borders :np.ndarray) -> None:
+        """
+        Setup the problem.
+
+        Args:
+            num_dim (int): number of dimensions of the function input.
+            num_obj (int): number of objectives (return values) of the function.
+            loss_function (callable): function to minimise.
+            borders (np.ndarray): the upper and lower values of the function domain.
+        """
+        super().__init__(
+            n_var = num_dim, 
+            n_obj = num_obj, 
+            xl = borders[0], 
+            xu = borders[1]
+        )
+        self.__loss_function = loss_function
+
+
+    def _evaluate(self, optim_array :np.ndarray[float], out :dict, *args :any, **kwargs :any) -> None:
+        """
+        Evaluates the loss function.
+
+        Args:
+            optim_array (np.ndarray[float]): the proposed input to the loss function.
+            out (dict): a pymoo dictionary to store constrains and results.
+        """
+        out['F'] = self.__loss_function(optim_array)
+
+
+
+
+class Optimiser():
+    """
+    This is an abstract class to define various kinds of optimsiers from.
+    """
+    def __init__(self, time_limit :str, algorithm :any) -> None:
+        """
+        Sets up the optimisers.
+
+        Args:
+            time_limit (str): the maximum time the optimsier should run for.
+            algorithm (any): the kind of optimiser.
+        """
+        self._limit = get_termination("time", time_limit)
+        self._algorithm = algorithm
+
+
+    def optimise(self, problem :Problem) -> any:
+        """
+        Minimises the problem.
+
+        Args:
+            problem (Problem): the problem describing the function to minimse.
+
+        Returns:
+            any: a result object containing the results and optimisation history.
+        """
+        res = minimize(
+            problem,
+            self._algorithm,
+            self._limit,
+            seed = 1,
+            save_history = True,
+            verbose = True
+        )
+        return res
+
+
+
+
+
+class NSGA2Optimiser(Optimiser):
+    """
+    MOO optimiser.
+    Simulates evolution.
```

Comment:
```suggestion
    Multi-objective optimiser. Simulates evolution.
```
Suggest removing the linebreak, ideally put more detail here too, i.e. you haven't defined what NSGA2 stands for, it would also be useful to be able to change the hyperparameters even if you include some nice defaults via kwargs